File: app/src/main/java/com/sogukj/pe/view/CustomViewPager.java
package com.sogukj.pe.view;

import android.content.Context;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;
import android.view.MotionEvent;

/**
 * Created by qinfei on 17/2/24.
 */

public class CustomViewPager extends ViewPager {

    int preX = -1;
    int preY = -1;

    public CustomViewPager(Context context, AttributeSet attrs) {
        super(context, attrs);

    }

    public CustomViewPager(Context context) {
        super(context);

    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent event) {

        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            preX = (int) event.getX();
            preY = (int) event.getY();
        } else {
           int dif= (int) (Math.abs((int) event.getX() - preX) - Math.abs(event.getY() - preY));
            if (dif>5) {
                return true;
            } else {
                preX = (int) event.getX();
                preY = (int) event.getY();
            }
        }
        return super.onInterceptTouchEvent(event);
    }
}

File: app/src/main/java/com/sogukj/pe/view/CustomViewPager.kt
package com.sogukj.pe.view

import android.content.Context
import android.support.v4.view.ViewPager
import android.util.AttributeSet
import android.view.MotionEvent

/**
 * Created by qinfei on 17/2/24.
 */

class CustomViewPager : ViewPager {

    internal var preX = -1
    internal var preY = -1

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {

    }

    constructor(context: Context) : super(context) {

    }

    override fun onInterceptTouchEvent(event: MotionEvent): Boolean {

        if (event.action == MotionEvent.ACTION_DOWN) {
            preX = event.x.toInt()
            preY = event.y.toInt()
        } else {
            val dif = (Math.abs(event.x.toInt() - preX) - Math.abs(event.y - preY)).toInt()
            if (dif > 5) {
                return true
            } else {
                preX = event.x.toInt()
                preY = event.y.toInt()
            }
        }
        return super.onInterceptTouchEvent(event)
    }
}